monix-eval/shared/src/test/scala/monix/eval/TaskLocalSuite.scala

```diff
@@ -268,4 +268,88 @@ object TaskLocalSuite extends SimpleTestSuite {
 
     t.runToFutureOpt
   }
+
+  testAsync("TaskLocal.isolate should properly isolate during async boundaries") {
+    implicit val s = TracingScheduler(Scheduler.singleThread("local-test"))
+      .withExecutionModel(ExecutionModel.AlwaysAsyncExecution)
+
+    val local = Local(0)
+    val test = for {
+      _ <- Task.evalAsync(local := 50)
+      _ <- TaskLocal.isolate {
+        Task.evalAsync(local := 100)
+      }
+      v <- Task.evalAsync(local())
+      _ <- Task.now(assertEquals(v, 50))
+    } yield ()
+
+    test.runToFutureOpt
+  }
+
+  testAsync("TaskLocal.isolate should properly isolate during async boundaries on error") {
+    implicit val s = TracingScheduler(Scheduler.singleThread("local-test"))
+      .withExecutionModel(ExecutionModel.AlwaysAsyncExecution)
+
+    val local = Local(0)
+    val test = for {
+      _ <- Task.evalAsync(local := 50)
+      _ <- TaskLocal.isolate {
+        Task.evalAsync(local := 100).flatMap(_ => Task.raiseError(DummyException("boom")))
+      }.attempt
+      v <- Task.evalAsync(local())
+      _ <- Task.now(assertEquals(v, 50))
+    } yield ()
+
+    test.runToFutureOpt
+  }
+
+  testAsync("TaskLocal.isolate should properly isolate during async boundaries on cancelation") {
+    implicit val s = TracingScheduler(Scheduler.singleThread("local-test"))
+      .withExecutionModel(ExecutionModel.AlwaysAsyncExecution)
+
+    val local = Local(0)
+    val test = for {
+      _ <- Task.evalAsync(local := 50)
+      _ <- TaskLocal.isolate {
+        Task.evalAsync(local := 100).start.flatMap(_.cancel)
+      }
+      v <- Task.evalAsync(local())
+      _ <- Task.now(assertEquals(v, 50))
+    } yield ()
+
+    test.runToFutureOpt
+  }
+
+  testAsync("Running task isolates locals") {
```

Comment:
This one fails too even for `runToFuture`, not sure why

Reply:
Small update:
```scala
val f = p.future

f.map(_ => {
    assert(local() == 0, s"received ${local()} != expected 0 in $method")
})
```

fails if `f` is not isolated

Reply:
@oleg-py Is it a bug or feature? :D

Reply:
I'd say it's a quirk, I don't think it can be solved.

Reply:
Yeah, I tried but failed miserably :D I guess we can live with it. I'm not sure if it's even a use case we should support